rfd/0124-plugin-secrets.md

```diff
@@ -0,0 +1,134 @@
+---
+authors: Michael Wilson (michael.wilson@goteleport.com)
+state: draft
+---
+
+# RFD 124 - Plugin static credentials
+
+### Required Approvers
+
+* Engineering @r0mant, @justinas
+* Security @reed
+* Product: @klizhentas
+
+## What
+
+Allow storage of plugin static credentials within Teleport in a safe and secure way.
+
+## Why
+
+As we grow our plugin support, we have discovered that the various services we
+hope to integrate do not offer a uniform mechanism for configuration. For instance,
+Slack uses oauth2, but Okta requires an API token, and Jamf requires a username and
+password.
+
+Today, the plugin houses a credentials object that allows for this sort of storage.
+However, unlike oauth tokens, static credentials pose a much greater risk of exposure,
+and rotating credentials can be a time consuming process.  To reduce the risks associated
+with storing static credentials within Teleport and to allow for easier credentials rotation,
+we would like to create a new resource within Teleport that is handled in an explicitly secure
+way.
+
+## Details
+
+### The `PluginStaticCredentials` object
+
+A new object will be needed to house static credentials. At present it will support:
+
+- API tokens
+- Usernames and passwords
+
+It should be noted that this object should *only* be readable by the auth service to prevent
+exfiltration of these credentials.
+
+```yaml
+kind: plugin_static_credentials
+version: v1
+metadata:
+  name: plugin-name
```

Comment:
It has to match the name of the plugin. I've added a comment to clarify this.

Reply:
@mdwn what if you change the name of the plugin? what if user creates a name of the plugin matching the secret they want to exfiltrate? 

What if we forget to change the name of the plugin? What if there are multiple API keys (during rotation) with different names.

Why not use the system annotations and match by annotation. For example, by default no plugin will be able to read the API key unless the annotations on the plugin and the API key match. System annotations can not be modified by users in any resource. You can have multiple API keys accessible to the plugin as long as annotations match.

Reply:
So as is currently described, the plugin credentials will be removed along with the plugin if the plugin is deleted, and it won't be possible to create static credentials without a corresponding plugin. So there shouldn't be this risk if having errant credentials in Teleport without an immediate need for them. Additionally, a user shouldn't be able to read the plugin credentials data anyway.

Right now we can't modify plugins, so plugins must be deleted/re-created. Because of the creation/deletion logic, that should also prevent leaking of plugin credentials.

Annotations are a possibility, and that's definitely a gap with this implementation -- it assumes only one type of credential per plugin. I'm not opposed to this -- it seems reasonable, and I was initially toying with the idea of multiple plugins wanting to access the same credential.

Is it sufficient here to do something like:

- `Plugin` object gets a `credentials annotations` field
- `PluginManager` queries the backend and provides anything with those annotations to the plugin on startup

Reply:
> Plugin object gets a credentials annotations field
> PluginManager queries the backend and provides anything with those annotations to the plugin on startup

Yes, assuming the value will match the annotation value and each plugin by default will get a unique annotation that is immutable and not editable by user.

I'd go annotations route, because this will diffuse potential limiting situation and I don't like the idea of binding the name that will be very hard to migrate from

Reply:
On further thought, one thing that worries me about the annotation approach is figuring out which credentials belong to which plugin. I can imagine a few situations here:

Say we've got two credentials objects which both match on annotations. Do we provide both? How does the plugin figure out which credentials it should use?

Say we have two Okta service plugins which both need API tokens.  In this case it's not sufficient to annotate the object `forPlugin: okta` or something like that, we'd have to say: `forPlugin:okta` and `pluginName: okta` .

IMO, it's more straightforward to have one credentials object per plugin because it eliminates potential problems with duplicate matching credentials or multiple copies of the same plugin.

Let me know if I'm off base here.

Reply:
FWIW I agree that matching by labels or annotations seems a better approach. This is consistent with how we do matching for other resources in Teleport as well, for example for dynamic app/db registration or in the discovery service.

Reply:
Ack, sounds good. I'll adjust!

Reply:
Okay, I've got the labels added in here. Let me know what you think.